### ISSUE TITLE
RFC: Multiprocess binaries and packaging options

### ISSUE THREAD

**Issue by ryanofsky** [bitcoin/bitcoin#30983]
Issue for discussion about ways multiprocess functionality could be packaged and released.

### Binaries

One goal of the [multiprocess project](https://github.com/bitcoin/bitcoin/issues/28722) has been to provide minimal binaries that only include node, wallet, or gui code, which spawn or connect to other processes as needed to provide other functionality.

The idea implemented in #10102 is to have 3 binaries:

- **`bitcoin-node`** - contains node code and libraries (leveldb), and no wallet or gui code
- **`bitcoin-wallet`** - contains wallet code and libraries (sqlite), and no node or gui code
- **`bitcoin-gui`** - contains gui code and libraries (qt), and no node or wallet code

Which are built differently than current binaries:

- **`bitcoind`** - contains node code+libraries (leveldb) and wallet code+libraries (sqlite)
- **`bitcoin-qt`** - contains node code+libraries (leveldb), wallet code+libraries (sqlite), and gui code+libraries (qt)

### Original release plan

My original idea implementing this was for there to be a separate multiprocess bitcoin release with the multiprocess binaries.

The main release would be unchanged, including existing `bitcoind` and `bitcoin-qt` binaries, but there would be a separate release containing `bitcoin-node`, `bitcoin-wallet` and `bitcoin-gui` binaries. The main release would be available in the normal place like *bitcoincore.org/bin/bitcoin-core-27.1/*, and the multiprocess release would be available alongside, at *bitcoincore.org/bin/bitcoin-core-multiprocess-27.1/* or someplace similar. Then, after a release cycle or two, the multiprocess release could replace the main release if it was working well.

But this approach has some drawbacks, namely that it would be a burden for maintainers to create separate releases and make them accessible, and that it could be confusing for users to encounter two releases with different binaries. So it doesn't seem like this approach is optimal (at least not for general users, maybe it would be ok for miners using the IPC mining interface).

### Release options overview

Taking a step back, it seems like there are 3 possible options for releasing multiprocess functionality:

1. **Side-release**: create a separate multiprocess release for multiprocess binaries, as described above.
2. **Side-binaries**: add multiprocess binaries (`bitcoin-node`, `bitcoin-wallet`, `bitcoin-gui`) to existing releases alongside existing binaries (`bitcoind` and `bitcoin-qt`).
3. **Combined binaries**: add multiprocess functionality to existing `bitcoind` and `bitcoin-qt` binaries that can be enabled/disabled with runtime arguments.

### Release options comparison

From discussion in recent PR's and offline, it seems like approach (3) might be favored, but it is worth considering pros and cons of the three approaches.

1. **Side-release**
   - Pros:
     - *Opt-in*: Keeps original binaries unchanged for users who don't want multiprocess functionality.
     - *Minimal binaries*: Node binary contains only node code not wallet code, gui binary contains no node or wallet code.
   - Cons:
     - *Extra work*: It's more work for maintainers and signers to create an extra set of packages and make them accessible.
     - *Complexity*: Posting a separate release could be confusing for users.
2. **Side-binaries**:
   - Pros (same as previous by being opt in and providing minimal binaries, but additionally):
     - *Easy to implement*: Implemented in [#30975](https://github.com/bitcoin/bitcoin/issues/30975) and basically requires no changes other than flipping cmake options.
   - Cons:
     - *Confusing*: This would probably be the most confusing approach for users. They would see one release with `bitcoind`, `bitcoin-qt`, `bitcoin-node`, `bitcoin-wallet`, `bitcoin-gui` binaries and not be able to easily understand how they are intended to be used.
   - Potential mitigations:
     - We would probably put multiprocess binaries in another folder (like `multiprocess/bin/` instead of `bin/`) so they are clearly labeled and aren't confused with the default binaries.
     - We might want to rethink our approach to packaging binaries to begin with. We are already shipping a substantial collection of binaries: `bitcoin-cli`, `bitcoind`, `bitcoin-qt`, `bitcoin-tx`, `bitcoin-util`, `bitcoin-wallet`, and `test_bitcoin` which users are expected to call individually, and are probably already somewhat confusing. By contrast, git ships dozens of binaries and nobody is confused by them because they are live in a libexec directory and are wrapped with single `git` command that finds and executes the right ones. Maybe taking `git` as inspiration we could have add a unified `bitcoin` command that looks like:

        ```bash
        bitcoin gui [OPTIONS]             # Start a gui
        bitcoin daemon [OPTIONS]          # Start a daemon
        bitcoin rpc [OPTIONS] COMMAND     # Call an rpc
        bitcoin wallet [OPTIONS] COMMAND  # Access a wallet
        bitcoin test [OPTIONS]            # Run tests
        bitcoin help
        # ... and more for bitcoin-util, bitcoin-tx ...

3. **Combined binaries**:
   - Pros:
     - *Compatibility*: Users could download and invoke `bitcoind` and `bitcoin-qt` binaries as before, not know they support IPC, not use IPC unless it's switched on by default, and there would be have runtime options to turn it on or off.
   - Cons:
     - *Bigger binaries and more dependencies*: Binaries will both contain node, wallet, and IPC code. There will be no non-IPC binaries users can deploy and run separately if they don't use IPC, and no node binary users can deploy and run separately if they don't use the wallet, or want to run the wallet and node on different hosts.
     - *Requires code changes*: Will require some thinking about how to add runtime options to [#10102](https://github.com/bitcoin/bitcoin/pull/10102) to control whether `bitcoind` should spawn separate wallet process or run wallet code internally, and similarly whether `bitcoin-gui` should use a separate node process or run node code in the same process. Will also require updating the build system.

As mentioned, approach (3) seems favored so far, though personally I like approach (2), especially if we could have a unified `bitcoin` command.

Creating this issue to get feedback and document any decisions we make.

**Comment by Sjors** [comment]
I like the idea of a unified `bitcoin` command.

We could start with seperate binaries for multiprocess, e.g.:

* `bitcoin daemon` (for `bitcoind`)
* `bitcoin daemon multiprocess` for `bitcoin-node`
* `bitcoin gui` (for `bitcoin-qt`)
* `bitcoin gui multiprocess` (for `bitcoin-gui`)

And perhaps not worry about just including multiprocess in the utilities:
* `bitcoin wallet` (for `bitcoin-wallet`)

Then later on the `multiprocess` could be made default and we drop the non-multiprocess binary.

The existing `bitcoind`, `bitcoin-qt` and `bitcoin-cli` and  binaries should probably stay where they are. Later on they can be changed to emit a deprecation warning "use bitcoin daemon instead", and then eventually deleted.

**Comment by hebasto** [comment]
> I like the idea of a unified `bitcoin` command.

So do I.

Would it be beneficial for users if we started shipping DEB and RPM packages? The package manager would handle the proper installation of multiple executables.

**Comment by ryanofsky** [comment]
re: https://github.com/bitcoin/bitcoin/issues/30983#issuecomment-2378828326

> We could start with seperate binaries for multiprocess, e.g.:

This makes sense. Initially I was thinking all the binaries had to live in the same place, but this is not true. We could leave all existing binaries where they are in `bin/` and just add 3 new files to releases when multiprocess option is turned on:

- `bin/bitcoin`
- `libexec/bitcoin-node`
- `libexec/bitcoin-gui`

Implementing this might be as simple as changing installation path for `bitcoin-node` and `bitcoin-gui` from `BINDIR` to `LIBEXECDIR` in cmake, and then writing the `bitcoin` wrapper as described above. The wrapper could just look for whatever binary it was trying to start in the `libexec/` directory first and `bin/` directory second.

On command lines you suggested, I think they make sense but might suggest:

- `bitcoin daemon` as synonym for `bitcoind`
- `bitcoin gui` as synonym for `bitcoin-qt`
- `bitcoin rpc` as synonym for `bitcoin-cli`
- `bitcoin -m daemon` as synonym for `bitcoin-node`
- `bitcoin -m gui` as synonym for `bitcoin-gui`

etc

re: https://github.com/bitcoin/bitcoin/issues/30983#issuecomment-2379073216

On DEB and RPM packages, I'm sure they could be useful for their own reasons, but I don't know if they offer things that would specifically help with this issue, and I think a `bitcoin` wrapper should be able to work well regardless of the release format.

**Comment by Sjors** [comment]
`bitcoin -m daemon -debug=net` as a format makes sense. It makes it more clear which arguments are passed into `bitcoin-node`.

**Comment by MrSuddenJoy** [comment]
@ryanofsky `bin/bitcoin` should (or could) be meta-process of all others that you mentioned. Just `symlink` maybe?

**Comment by ryanofsky** [comment]
> @ryanofsky bin/bitcoin should (or could) be meta-process of all others that you mentioned. Just symlink maybe?

It could be a symlink to existing executable, which could interpret its arguments as needed based on argv[0], instead of being a new executable. But I think having a new executable that calls exec https://man7.org/linux/man-pages/man3/exec.3.html would be more straightforward

**Comment by willcl-ark** [comment]
I very much like the idea of the wrapper approach described in https://github.com/bitcoin/bitcoin/issues/30983#issuecomment-237920497, it feels to me like the correct way to structure and package an application consisting of multiple binaries rather than overloading existing binaries with even more options.

My only reservation with the approach is the potential magnitude of the effects downstream. If executed poorly this could break many scripts, programs and general infra. However, so long as we maintain the current binary names in the current directory, as currently proposed, and only introduce a new wrapper around them, then it seems like it should mitigate any downstream impact (completely).

**Comment by ryanofsky** [comment]
Here's an initial implementation of a `bitcoin` wrapper executable: commit 65e49f4579ee823284bf04f1295b59116acc727b [(branch)](https://github.com/ryanofsky/bitcoin/commits/pr/wrap)

This should be purely additive, and to willcl-ark's point, not break backwards compatibility. It just lets us add new binaries to a libexec/ directory instead of bin/.

**Comment by MrSuddenJoy** [comment]
@ryanofsky Not entirely sure this is way to go imho.

As long as everything works well ( = as expected ) its all ok, but as soon as things go south, problems (especially ones that are hard to solve/complex) start to accumulate/pile up. And thats exactly whats root cause of many bad things, namely frustration, burn-out and things like this.

Having `meta-package` has proven (to me) to be perfect cure for problems described above. If something breaks, its dead simple to: localise where (in code) fault is, and fix it.

**Comment by josibake** [comment]
Echoing others, I like option 2 (side-binaries). A `bitcoin <cmd>` wrapper on top of multiple binaries also seems quite nice. While this might cause some confusion for end users, I think this will be far less confusing than having a side release (and much less work for maintainers and guix builders).

Regarding option 3, one of the value propositions of multiprocess in my mind is being able to run a binary for a specific purpose, with only the code needed for that purpose in the binary, e.g., running `bitcoin-node` on a server without wanting/needing a wallet or gui. Option 3 seems to undermine that, while also introducing complexity (as mentioned in the cons).

**Comment by ryanofsky** [comment]
The `bitcoin` wrapper discussed above is added in PR #31375

**Comment by darosior** [comment]
I don't think the cons of option 3 are necessary to get its pros. How about simply keeping `bitcoind` and `bitcoin-qt` which would start `-node`, `-wallet` and `-gui` binaries from `libexec`? The bigger binaries are a consequence of wanting to be able to disable IPC through a runtime option. It's an internal, why is it something we would want to make possible for the user to change? And why should we keep releasing the large binaries? If we do, when do we stop? I think that *once we are confident* of shipping multiprocess-enabled binaries we should just do it and not stay in a halfway situation.

**Comment by ryanofsky** [comment]
re: https://github.com/bitcoin/bitcoin/issues/30983#issuecomment-2620155964

> I don't think the cons of option 3 are necessary to get its pros. How about simply keeping `bitcoind` and `bitcoin-qt` which would start `-node`, `-wallet` and `-gui` binaries from `libexec`? 

Nice idea! This is an interesting hybrid of approaches 2 & 3 and I'll add it as a 4th option in the description.

4. **Minimally combined binaries**: add limited multiprocess functionality to existing `bitcoind` and `bitcoin-qt` binaries that can be enabled/disabled with runtime arguments. Difference from option 3 is that IPC side-binaries would also exist and `bitcoind` and `bitcoin-qt` would not contain IPC code, but automatically `exec()` IPC binaries them `-ipcbind` or other IPC options were enabled.

4. **Minimally combined binaries**:
   - Pros:
     - Combines pros listed in option 1 with pros listed in option 3.
   - Cons:
     - *Some implementation messiness*: A downside of this approach is options will need to be parsed twice. For example `bitcoind` will need to read arguments and config file to detect if any ipc features were requested, then it will need to exec `bitcoin-node` which will read the same options and configuration again. In general this doesn't seem as nice and clean as having a single wrapper binary like #31375 that calls all the other binaries, and can show help displaying what subcommands are available.

---

> The bigger binaries are a consequence of wanting to be able to disable IPC through a runtime option. It's an internal, why is it something we would want to make possible for the user to change? And why should we keep releasing the large binaries? If we do, when do we stop? I think that _once we are confident_ of shipping multiprocess-enabled binaries we should just do it and not stay in a halfway situation.

I think my brain might be a little fried, but I am not following this part. Maybe you can be more concrete and say what negative outcome(s) you would like to avoid?

**Comment by darosior** [comment]
> Maybe you can be more concrete and say what negative outcome(s) you would like to avoid?

Sure, i meant i don't see why we should release non-IPC binaries once we have decided to enable multiprocess. It seems it would make us be in this halfway situation of keeping around the bigger binaries with no clear reason nor timeline about when we'd get rid of them.

**Comment by ryanofsky** [comment]
> Sure, i meant i don't see why we should release non-IPC binaries once we have decided to enable multiprocess.

Got it. I honestly don't have an opinion about that. I think existing non-ipc binaries should be kept as long as features like adding a Chain IPC interface (#29409) or removing wallet code from `bitcoin-node` process (#10102) are planned. But after that point you could easily drop them, especially if #31375 gained adoption. You could also remove them in phases, like (1) move them to libexec (2) change `bitcoin` command default to run multirprocess instead of monolithic binaries (3) remove monolithic binaries.

**Comment by darosior** [comment]
I agree the multiprocess project should be finished before we consider enabling it and shipping binaries. I guess that ties into the discussion in #31756 so i'll expand there.

**Comment by MrSuddenJoy** [comment]
Just bear in mind that bundled version means slower installation/operation time as well as higher bandwitch usage.

**Comment by Sjors** [comment]
The wrapper binary has been merged with #31375. Actually shipping the binaries would happen if #31802 lands.

**Comment by maflcko** [comment]
> The wrapper binary has been merged with [#31375](https://github.com/bitcoin/bitcoin/pull/31375). Actually shipping the binaries would happen if [#31802](https://github.com/bitcoin/bitcoin/pull/31802) lands.

This was merged yesterday.

**Comment by ryanofsky** [comment]
sipa, theuni, TheBlueMatt, and others raised a question in #31802 that wasn't directly considered here and is still relevant: **Should `bitcoind` accept an `-ipcbind` option?**

### Background

The multiprocess code and libmultiprocess/cap'nproto dependencies allow for two distinct features:

1. An *IPC listening* feature (`-ipcbind`), which lets the bitcoin node listen on a Unix socket and accept IPC connections from utilities and third-party applications.
2. A *process separation* feature, which allows node/GUI/wallet code to be built and run in separate processes, as described in this issue, and would also let Bitcoin Core spawn and communicate with processes that do other things like indexing.

### Objections in 31802

The comments in #31802 pointed out that if we were **only** going to implement and enable the *IPC listening* feature (1), and never implement the *process separation* feature (2), the release process chosen and discussed here, implemented in #31375 and #31679, and toggled on in #31802, would not make much sense. In that case, it would be simpler to add an `-ipcbind` option to the existing `bitcoind` and `bitcoin-qt` binaries, rather than introducing a new `bitcoin` wrapper with `-m` and `-M` options.

I think this objection is mostly *procedural*[^1]. I think sipa and theuni would say that after it became clear feature (1) was gaining more traction than feature (2), I should have reconsidered the release process and streamlined it to support only feature (1). This would mean probably not introducing the `bitcoin` wrapper, and definitely not giving it `-m`/`-M` options. This is a fair take, and if you have short term and long term goals it doesn't always make sense to optimize for the long term goal, especially when there is uncertainly around it.

But the objection is not just procedural; it also has some practical implications:

1. The current code leaves `bitcoind` and `bitcoin-qt` without `-ipcbind` support. The feature is only available through the new `bitcoin` command.
2. The current code requires users who want to use `-ipcbind` to explicitly choose the `-m` option and opt into multiprocess binaries. It's not good to force users to understand an internal implementation detail like this, and #33229 fixes this.
3. The current code adds new files to the release (`bin/bitcoin` and `libexec/*` binaries), which users may find confusing or unfamiliar.
4. Theoretically, if/when process separation features roll out, sipa was worried that people who ["want the mining IPC but don't care about multiprocess"](https://github.com/bitcoin/bitcoin/pull/31802#issuecomment-3206057126) might be forced into using process separation they never asked for. IMO, this should never happen and would clearly be a mistake if it did. The GUI/node/wallet process separation implemented in #10102 should be controlled by a different runtime flag, independent of `-ipcbind`[^2].

### Should `bitcoind` and `bitcoin-qt` accept `-ipcbind`?

I think the answer should be no for now. I think it's good for now that these binaries don't include cap'nproto, and in general (especially after events like the xz exploit) it is a positive thing if we structure our build in a way that minimizes unnecessary dependencies when it's reasonable and doesn't impose significant costs. I also think it's good to gradually encourage more use and familiarity with the `bitcoin` command, and to normalize the idea that while existing `bitcoind` and `bitcoin-cli` commands will continue to work and remain backwards compatible, some new features will only be available through the `bitcoin` command, particularly IPC-related features.

The situation could change in the future. If we roll out other IPC features beyond the mining interface (#32297, #29409, #32898), we might want to make them available in `bitcoind`. Users may ask for this if they prefer not to use the `bitcoin` command. We may also want to reduce the number of binaries to shrink release size[^3]. I would want to wait for a more concrete reason to do this, but do think it's a reasonable option to consider, especially going forward.

[^1]: In the PR I called this a "symbolic" objection, but "procedural" should be more accurate.  
[^2]: There are practical reasons process separation requires a runtime option. The IPC interfaces between node, wallet, and GUI have a large surface area, and enabling process separation has wide-ranging effects. The feature can't just be enabled with a one-time flag flip (look up "firefox electrolysis" for an analogue). A practical rollout would first make the feature available at runtime to users but off by default, then on by default with a runtime switch to turn it off, and only later with the runtime switch removed.  
[^3]: sipa has also raised concern that a larger number of binaries risks "splitting attention" and reducing testing quality. That concern would be more applicable if the binaries contained different code, or code compiled with different preprocessor defines or compiler options. But in this case, the binaries all contain the same compiled objects, just linked differently. As a thought experiment, imagine a typical Linux system with many binaries, versus the same system with all code linked into a single binary. Test coverage from running code would be exactly the same in both cases. Linking unused code into a binary doesn't provide meaningful test coverage. It's true that changing how code is linked can *indirectly* increase test coverage by preventing certain runtime option combinations, but you could achieve the same effect by directly disallowing those option combinations.

**Comment by Sjors** [comment]
Another thing to consider, in the context of binary size, is if we want to drop `bitcoin-gui` for now, as I don't think it's very common for anyone to be connecting a miner to a GUI (though maybe Windows miners would?)

Speaking of Windows, if we do #33229 then I think the Windows desktop shortcut should use the `bitcoin` binary instead of `bitcoin-qt`.

**Comment by ryanofsky** [comment]
There will probably be more PRs implementing ideas discussed here, but they can be worked on in separate PR's and issues and this can be closed. Possible followups:

1. Adding `-ipcbind` option to `bitcoind` to make IPC features available without using `bitcoin` command and to reduce number of internal binaries that need to be built and installed [(comment)](https://github.com/bitcoin/bitcoin/issues/30983#issuecomment-3210296557).

2. Implementing `-ipcnode` and `-ipcwallet` options (probably as part of [#10102](https://github.com/bitcoin/bitcoin/pull/10102)) choosing whether to spawn the node from the gui, and the wallet from the node or gui as separate processes, instead of together in one process. This could let process separation be released first as optional and off by default, then optional and on by default, then as required [(comment)](https://github.com/bitcoin/bitcoin/issues/30983#issuecomment-3210296557).

3. Maybe dropping `bitcoin-gui` binary and not supporting IPC connections to the GUI to reduce number of binaries [(comment)](https://github.com/bitcoin/bitcoin/issues/30983#issuecomment-3210606461).

4. Updating deskop shortcuts to use `bitcoin gui` instead of `bitcoin-qt` [(comment)](https://github.com/bitcoin/bitcoin/issues/30983#issuecomment-3210606461).
